### ISSUE TITLE
kubectl does not work with namespace authorization.

### ISSUE THREAD

**Issue by f-higashi** [kubernetes/kubernetes#13097]
I created policy.jsonl
```
{"user":"alice", "namespace": "projectCaribou"}
```

And I tried to get pods by kubectl.
```
$ kubectl get po --username=alice --password=password --server=https://localhost:6443 --insecure-skip-tls-verify=true
error: couldn't read version from server: the server does not allow access to the requested resource
```
At first, kubectl seems to access https://localhost:6443/api/ and fail because the "alice" can only have access on https://localhost:6443/api/v1/namespaces/projectCaribou.

**Comment by mikedanese** [comment]
pass `--namespace=projectCaribou` with your kubectl command

**Comment by f-higashi** [comment]
I tried it. But the same error occurs.

**Comment by f-higashi** [comment]
I added detail below.

```
KUBE_API_ARGS="--v=5 --basic_auth_file=/home/vagrant/basic_auth.csv \
               --authorization_mode=ABAC --authorization_policy_file=/home/vagrant/auth_policy.jsonl"
kube-apiserver  $KUBE_API_ARGS --service-cluster-ip-range=10.0.0.0/16 --etcd_servers=http://127.0.0.1:4001 
```

basci_auth.csv
```
admin,admin,admin
password,alice,alice
```
auth_policy.jsonl
```
{"user":"admin"}
{"user":"alice","namespace":"projectCaribou"}
```

```
vagrant@vagrant-ubuntu-trusty-64:~$ kubectl get po --username=alice --password=password --server=https://localhost:6443 --insecure-skip-tls-verify=true --namespace=projectCaribou
error: couldn't read version from server: the server does not allow access to the requested resource
vagrant@vagrant-ubuntu-trusty-64:~$ kubectl get po --username=admin --password=admin --server=https://localhost:6443 --insecure-skip-tls-verify=true --namespace=projectCaribou
NAME      READY     STATUS    RESTARTS   AGE
```

When I modify auth_policy.jsonl such bellow,  kubectl command is success. but off course alice can access other namespaces.
```
{"user":"admin"}
{"user":"alice","resource":""}
{"user":"alice","namespace":"projectCaribou"}
```

in success case, kube-apiserver output bellow 

```
I0825 07:41:49.412858    7385 handlers.go:142] GET /api: (280.456µs) 200 [[kubectl/v1.1.0 (linux/amd64) kubernetes/6630932] 127.0.0.1:45213]
I0825 07:41:49.414010    7385 handlers.go:142] GET /api/v1/namespaces/projectCariboua/pods: (669.725µs) 200 [[kubectl/v1.1.0 (linux/amd64) kuber
```
in fail case, kube-apiserver output only one line

```
I0825 07:47:06.181911    7486 handlers.go:142] GET /api: (104.081µs) 403 [[kubectl/v1.1.0 (linux/amd64) kubernetes/6630932] 127.0.0.1:45218]
```

**Comment by elsonrodriguez** [comment]
Having a similar issue with 1.0.1, but abac code hasn't changed for a while. Whenever namespace is restricted via ABAC policy, all calls with kubectl just fail.

This may also have gone un-noticed because in parts of the documentation the namespace key is said to be "ns", which is not read by the abac code, so people switching to the "ns" syntax vs "namespace" see success, but they don't test to see if they can perform actions on other namespaces.

This may be a duplicate of #12107 and may be a related to #12495.

**Comment by f-higashi** [comment]
I checked source code of kubectl.
I think it is the fail reason that kubectl calls client.NegotiateVersion(&config, version) 
https://github.com/kubernetes/kubernetes/commit/f31191224bd9b555839e3aef2a49cbb3ccb73f69#diff-11e9bf777cb7b0498118f497a570f50cR61

from this commit (v0.20~), kubectl try to get version information and fail.

So It is needed to remove negotiating verion in kubectl or to allow authenticated user to access some endpoints like /api/, /version.

**Comment by bgrant0607-nocc** [comment]
cc @kubernetes/kubectl @kubernetes/rh-platform-management

**Comment by liggitt** [comment]
for an authorizer to able to express "allow all users to access /api and /version", the authorizer attributes need to include the request path for non-resource requests. @deads2k, might be worth adding non-resource-path to the attributes in https://github.com/kubernetes/kubernetes/pull/12209/files#diff-a80bc00f0f154458f2f259c2e7e61e52R13

**Comment by liggitt** [comment]
the alternative is to leave those endpoints outside the authorizer filter, but I prefer having the option to restrict access (/healthz might be considered sensitive info by some admins, for example, while others wouldn't care who saw it)

**Comment by mkulke** [comment]
Is there a fix/workaround for 1.0 conceivable, or is this an architectural issue? i'd like to give users access to specific namespaces via kubectl w/o exposing everything.

**Comment by mikedanese** [comment]
We should fix this before v1.1

**Comment by davidopp** [comment]
@smarterclayton @liggitt @eparis  is there anyone at RH who might be able to do this for 1.1?

**Comment by smarterclayton** [comment]
I think this is blocked on the other issue (authorizer attributes and extended data) that @deads2k owns - @erictune had final review on that but was concerned about putting it into 1.1 (having time to review it).  Based on everything we've said here I'm a bit worried that we'd be rushing things we've been hesitant to rush.  We could add a special hardcoded case for /api/, but I am extremely hesitant to do that (we'd have to also do /apis/, and then probably /swaggerapi/, and ...)

**Comment by liggitt** [comment]
I don't want to hardcode allowed paths, since the version negotiation and validation stuff is very much in active development... I think https://github.com/kubernetes/kubernetes/issues/13667 is close... trying to get back to it today

**Comment by mkulke** [comment]
`/api` is not sufficient for create/update operations, as they also match their objects against the swagger endpoints to check validity. So you'd have to hardcode at least those and maybe more.

**Comment by davidopp** [comment]
@smarterclayton can you say something about the priority of this? We have it listed as v1.1 but that seems unlikely.
